src/d3d9/d3d9_fixed_function.cpp

```diff
@@ -1115,6 +1115,7 @@ namespace dxvk {
         default:
         case (DXVK_TSS_TCI_PASSTHRU >> TCIOffset):
           transformed = m_vs.in.TEXCOORD[inputIndex & 0xFF];
+          count = 4;
```

Comment:
This is wrong, `TCI_PASSTHRU` doesn't set `count = 4` because it's meant to use the size of the texcoord in the vertex decl.

Is this required to close the other issue? If so I imagine there is something else going on, but this is definitely not the right solution!

https://learn.microsoft.com/en-us/windows/win32/direct3d9/d3dtexturetransformflags

Reply:
Are you hitting a case here where `count == 0`? Curious... That might be the `D3DTTFF_DISABLE` case that is missing some handling.

Reply:
Not really. I'm hitting `count == 7.` Which is more than texcoord size

Reply:
For case if we have `count` smaller than 4 can suggest this:
```suggestion
          count = std::min(count, static_cast<uint32_t>(4));
```

Reply:
one more way it to prevent `D3DTTFF_DISABLE` if we have `count > 4` [here](https://github.com/doitsujin/dxvk/blob/101617753e9d4279877b9f7e0fa0721ffd6b8083/src/d3d9/d3d9_fixed_function.cpp#L1173). Obviously, it doesn't make sense to do `opCompositeExtract` if we don't `opCompositeInsert`(which won't do if count > 4)

Reply:
BTW, I'm not pretty sure it's a good way to get `count` assigning it to bit mask
https://github.com/doitsujin/dxvk/blob/101617753e9d4279877b9f7e0fa0721ffd6b8083/src/d3d9/d3d9_fixed_function.cpp#L1113
It looks for me that correct way to get count from bit mask is sth like this:
```
count = 0;
for (int i = 0; i<3; ++i)
  if ((flags >> i) & 1)
  	++count;
```

Reply:
It's not like that though, the 'flags' are actually just a regular value.

```cpp
typedef enum _D3DTEXTURETRANSFORMFLAGS {
    D3DTTFF_DISABLE         = 0,    // texture coordinates are passed directly
    D3DTTFF_COUNT1          = 1,    // rasterizer should expect 1-D texture coords
    D3DTTFF_COUNT2          = 2,    // rasterizer should expect 2-D texture coords
    D3DTTFF_COUNT3          = 3,    // rasterizer should expect 3-D texture coords
    D3DTTFF_COUNT4          = 4,    // rasterizer should expect 4-D texture coords
    D3DTTFF_PROJECTED       = 256,  // texcoords to be divided by COUNTth element
    D3DTTFF_FORCE_DWORD     = 0x7fffffff,
} D3DTEXTURETRANSFORMFLAGS;
```

Reply:
What game are you seeing this with btw? Is there an apitrace available?

Maybe we need to consider the vertex declaration's TEXCOORD stuff or something too, or just clamp it... Not sure.

Reply:
https://github.com/doitsujin/dxvk/issues/3293#issuecomment-1475057852

Reply:
Ya I see this in the trace:
`2031784 @0 IDirect3DDevice9::SetTextureStageState(this = 0xca51c8, Stage = 3, Type = D3DTSS_TEXTURETRANSFORMFLAGS, Value = -1) = D3D_OK`

Reply:
(Obviously you are seeing 7 because of the & with 0b111 -> 7, just to be clear :p)

Reply:
We already handle the vertex decl thing later on with:

```cpp
          for (uint32_t j = count; j < 4; j++) {
            // If we're outside the component count of the vertex decl for this texcoord then we pad with zeroes.
            // Otherwise, pad with ones.

            // Very weird quirk in order to get texcoord transforms to work like they do in native.
            // In future, maybe we could sort this out properly by chopping matrices of different sizes, but thats
            // a project for another day.
            uint32_t texcoordCount = (m_vsKey.Data.Contents.TexcoordDeclMask >> (3 * inputIndex)) & 0x7;
            uint32_t value = j > texcoordCount ? m_module.constf32(0) : m_module.constf32(1);
            transformed = m_module.opCompositeInsert(m_vec4Type, value, transformed, 1, &j);
          }
```

So I am happy for us to just do a `count = min(count, 4)` here for the solution.

Reply:
Once you have done that I will merge it.

Reply:
Hmmm now I'm wondering if that should really be `count = min(count, whatever to read from texcoord deck mask)` actually given how that for loop works.

Reply:
I suggest  https://github.com/doitsujin/dxvk/blob/101617753e9d4279877b9f7e0fa0721ffd6b8083/src/d3d9/d3d9_fixed_function.cpp#L1173
`if (type != D3DTTFF_DISABLE && count < 4) {`
the routine below doesn't make any sense when count is 4 or bigger. the only thing we do is `opCompositeExtract` (in the way that makes spirv invalid)and don't use the value

Reply:
We still need to do the matrix tho

Reply:
then, we should rewrite for cicles to

Reply:
This is what I mean https://github.com/doitsujin/dxvk/tree/d3d9_fixed_function_tci

Reply:
Okay. Do you want me to grab the commit?
https://github.com/doitsujin/dxvk/commit/9d88e38c016aaac4c84882c208d1457884c63b52I 

With the `for` routine I meant we construct  transform only with `std::array<uint32_t, 4>`.
There is no sense in setting indeces from 4th.

Reply:
Ya. Can you make sure it fixes the issue for you as well?

Reply:
How can we have `texcoordCount` > 4 though? Isn't it literally from the decl. What's going on there? I am travelling now for a bit so won't get a chance to look myself.

Reply:
`count == 7 && texcoordCount == 0`
Good for you ;)

Reply:
If texcoordCount == 0, we probably want to treat it as `NINED3DTSS_TCI_DISABLE` right?

Reply:
Sry, what is `NINED3DTSS_TCI_DISABLE `?
are you talking about [this ](https://github.com/doitsujin/dxvk/pull/3390/files#diff-f8e71b8e89bd998acbde4d82f201fa4f9cd73dfc7432eb2d905525df7bf476abR1182)?

Reply:
Yeah, I had a look at what Gallium Nine (see link above) did and it disabled it in that case. So I imagine that's probably what we should do.

Reply:
Updated. Pls, check if it's what you meant :)